### PR TITLE
RELATED: ONE-3392 Color mapping

### PR DESCRIPTION
**Pull request by peterlacko**

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1825/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1310/ : FAILURE in 5m 35s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1829/ : SUCCESS in 18s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1314/ : FAILURE in 2m 59s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1832/ : SUCCESS in 1m 08s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1317/ : FAILURE in 5m 08s

File: src/components/visualizations/chart/Chart.tsx [hunk 1]
```diff
@@ -39,9 +39,31 @@ export interface IChartLimits {
     dataPoints?: number;
 }
 
+export interface IPaletteColor {
+    type: string;
+    value: string;
+}
+
+export interface IRGBColor {
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/Chart.tsx]: reuse it also in IColorPaletteItem

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1846/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1326/ : SUCCESS in 1m 49s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1848/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1328/ : SUCCESS in 2m 57s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1863/ : SUCCESS in 20s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1341/ : FAILURE in 5m 13s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1869/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1347/ : FAILURE in 5m 47s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1874/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1356/ : FAILURE in 6m 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1876/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1358/ : SUCCESS in 9m 09s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1877/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1359/ : FAILURE in 3m 27s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1879/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1361/ : FAILURE in 11m 05s

File: src/components/visualizations/chart/colorFactory.ts [hunk 2]
```diff
@@ -39,98 +46,184 @@ export const attributeChartSupportedTypes = [
...
+        const colorMap = colorMapping.find(item => item.id === itemId);
+        if (colorMap === undefined) {
+            return undefined;
+        }
+
+        if (colorMap.color.type === 'guid') {
+            return colorPalette.find(colorPaletteItem => colorPaletteItem.guid === colorMap.color.value);
+        } else {
+            return {
+                guid: String(Math.floor(Math.random() * 10000)),
```
Comment (pbenes) [inline src/components/visualizations/chart/colorFactory.ts]: why is this needed?

File: src/components/visualizations/chart/colorFactory.ts [hunk 3]
```diff
@@ -39,98 +46,184 @@ export const attributeChartSupportedTypes = [
...
+                color: getRgbString(colorConfig.color)
+            };
         });
     }
 }
 
+function getAttributeColorMapping(attribute: any, colorPalette: IColorPalette, colorMapping: IColorMap[]) {
+    const itemsCount = attribute.items.length;
+    let currentColorPaletteIndex = 0;
+    return range(itemsCount).map((itemIndex) => {
```
Comment (pbenes) [inline src/components/visualizations/chart/colorFactory.ts]: just wondering if we could use `attribute.items.map((item, index) => { ... })`;

File: src/components/visualizations/chart/colorFactory.ts [hunk 4]
```diff
@@ -39,98 +46,184 @@ export const attributeChartSupportedTypes = [
...
+            }
+        }
+
         if (colorPalette && isCustomPalette(colorPalette) && colorPalette[0]) {
-            return this.getCustomHeatmapColorPalette(colorPalette[0]);
+            return this.getCustomHeatmapColorPalette(colorPalette[0]).map(color => ({ color }));
         }
-        return HEATMAP_BLUE_COLOR_PALETTE;
+
+        return HEATMAP_BLUE_COLOR_PALETTE.map((color) => {
```
Comment (pbenes) [inline src/components/visualizations/chart/colorFactory.ts]: maybe we could keep the same styling as above, i.e. `.map(color => ({ color }));`

File: src/components/visualizations/chart/test/chartOptionsBuilder.spec.ts [hunk 5]
```diff
@@ -1812,22 +1834,25 @@ describe('chartOptionsBuilder', () => {
...
+            const attributeColorStrategy = new AttributeColorStrategy(
                 DEFAULT_COLOR_PALETTE,
+                undefined,
                 mVS[0],
                 mVS[1],
                 mVS[2],
                 fixtures.bubbleChartWith3MetricsAndAttribute.executionRequest.afm
             );
 
+            // console.log('measureStrategy', metricColorStrategy.getColorMapping());
```
Comment (pbenes) [inline src/components/visualizations/chart/test/chartOptionsBuilder.spec.ts]: remove log

File: stories/core_components/Heatmap.tsx [hunk 6]
```diff
@@ -332,4 +332,30 @@ storiesOf('Core components/Heatmap', module)
...
+                <Heatmap
+                    projectId="storybook"
+                    measure={MEASURE_1}
+                    columns={ATTRIBUTE_1}
+                    rows={ATTRIBUTE_2}
+                    onError={onErrorHandler}
+                    LoadingComponent={null}
+                    ErrorComponent={null}
+                    config={{
+                        // ...CUSTOM_COLOR_PALETTE_CONFIG,
```
Comment (pbenes) [inline stories/core_components/Heatmap.tsx]: maybe remove this commented line if we want to use default palette?

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1883/ : SUCCESS in 36s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1371/ : FAILURE in 4m 38s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1884/ : SUCCESS in 30s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1372/ : SUCCESS in 4m 14s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1889/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1377/ : FAILURE in 7m 32s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1894/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1382/ : FAILURE in 49s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1895/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1383/ : FAILURE in 46s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1896/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1384/ : FAILURE in 45s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1897/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1385/ : FAILURE in 44s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1899/ : SUCCESS in 20s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1386/ : SUCCESS in 6m 40s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1907/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1392/ : SUCCESS in 2m 58s

**Comment by lupko** [comment]
@peterlacko please put this PR on hold; we need to go over this design together - I have some reservations at the moment and would favor a more flexible, predicate based solution. 

e.g. mapping = list of { predicate: fun(), color: '...'}; if predicate evaluates true for an item being visualized, then the color is used; the catch here being that the predicate function does not have to be coded from scratch all the time: we would have a set of pre-defined predicates (type match, value match, uri match etc) that the users can exploit out of the box. furthermore, for SDK users we would have a simple factory fun to construct the predicates & even the mapping in a simple fashion.

let's talk more tomorrow.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1914/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1400/ : SUCCESS in 2m 38s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1933/ : SUCCESS in 23s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1422/ : FAILURE in 8m 24s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1943/ : SUCCESS in 1m 17s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1427/ : FAILURE in 4m 17s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1945/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1429/ : FAILURE in 59s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1947/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1431/ : FAILURE in 5m 59s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1949/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1433/ : FAILURE in 1m 01s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1952/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1436/ : SUCCESS in 2m 28s

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by peterlacko** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1988/ : SUCCESS in 20s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1468/ : SUCCESS in 11m 47s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/811/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1989/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/440/ : UNSTABLE in 17m 45s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1994/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1471/ : FAILURE in 1m 40s

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1996/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1472/ : FAILURE in 2m 24s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1997/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1473/ : FAILURE in 2m 28s

**Comment by peterlacko** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1998/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1474/ : SUCCESS in 2m 30s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/813/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1999/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/444/ : UNSTABLE in 17m 03s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by peterlacko** [comment]
extended test - cafe

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2004/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/446/ : FAILURE in 16s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2003/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1477/ : SUCCESS in 1m 57s

**Comment by peterlacko** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/814/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2005/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/447/ : SUCCESS in 17m 41s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2017/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1485/ : SUCCESS in 11m 28s

File: src/components/visualizations/chart/colorFactory.ts [hunk 7]
```diff
@@ -39,110 +55,287 @@ export const attributeChartSupportedTypes = [
...
+    protected createPalette(
+            colorPalette: IColorPalette,
+            colorAssignment: IColorAssignment[],
+            _viewByAttribute: any,
+            _stackByAttribute: any):
+        string[] {
+        return colorAssignment.map((map) => {
+            const color = isGuidColorItem(map.color)
+                ? getColorByGuid(colorPalette, map.color.value) : map.color.value;
+            return getRgbStringFromRGB(color);
```
Comment (ZdenekVidensky) [inline src/components/visualizations/chart/colorFactory.ts]: What if color is null?

Comment (peterlacko) [reply src/components/visualizations/chart/colorFactory.ts]: fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2019/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1488/ : SUCCESS in 5m 50s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2020/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1489/ : SUCCESS in 11m 58s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/465/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1491/ : SUCCESS in 1m 51s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/820/ : SUCCESS in 17m 16s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/406/ : SUCCESS in 22m 12s

**Pull request merged**